### PR TITLE
fixing flow sensor flow rate calibration issue

### PR DESCRIPTION
**Pull request by tongju12**
## Description
MR3K4 flow sensor GUI reading was wrong due to the wrong calibration from raw bit to flow rate. It showed 15 lpm while ME said the real flow rate should be around 2 lpm.

Jing and I checked the datasheet of EL3054 and recalibrate from the raw bit to match the flow capacity from the flow sensor DPM. After that, it shows 1.8 lpm which makes sense. The old calibration was wrong.
## Motivation and Context

The reading of flow sensor is not correct according to the limit of flow sensor capacity (0-5 lpm)
Here is the old FB : https://github.com/pcdshub/lcls-twincat-general/pull/82
Recalibrate the formula by the datasheet and manually calibrate first to get reasonable value.

https://jira.slac.stanford.edu/browse/ECS-3783
## How Has This Been Tested?

After I log in with the new FB, the current flow sensor reading of MR3K4 is 1.8 lpm and consistent with ME estimation 2lpm.

## Where Has This Been Documented?

https://jira.slac.stanford.edu/browse/ECS-3783)

## Pre-merge checklist
- [ ] Code works interactively
- [ ] Code contains descriptive comments
- [ ] Test suite passes locally
- [ ] Libraries are set to fixed versions and not ``Always Newest``
- [ ] Code committed with ``pre-commit`` (alternatively ``pre-commit run --all-files``)

### CODE REVIEW

**Comment by tongju12** [comment]
I just run pre-commit run --all-files to fix those errors.

File: lcls-plc-tmo-optics/tmo_optics/POUs/FB_AnalogInput1.TcPOU [hunk 1]
```diff
@@ -0,0 +1,36 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<TcPlcObject Version="1.1.0.1" ProductVersion="3.1.4022.18">
+  <POU Name="FB_AnalogInput1" Id="{96db7639-6b01-4d7a-bcbc-8bfc0195f2f2}" SpecialFunc="None">
+    <Declaration><![CDATA[FUNCTION_BLOCK FB_AnalogInput1
+VAR_INPUT
+    // Connect this input to the terminal
+    iRaw AT %I*: INT;
+    ratio: REAL;
```
Comment (ZLLentz) [inline lcls-plc-tmo-optics/tmo_optics/POUs/FB_AnalogInput1.TcPOU:8]: this should not be an input parameter

File: lcls-plc-tmo-optics/tmo_optics/POUs/FB_AnalogInput1.TcPOU [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    // The real flow rate read from the output
+    fReal: REAL;
+END_VAR
+VAR
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[   IF (iTermBits=0) THEN iTermBits := 32767; END_IF
+   ratio := INT_TO_REAL(iRaw)/UINT_TO_REAL(iTermBits);
+   rC := 16.0*ratio + 4.0; // transfer the current mA base on raw bit reading
```
Comment (ZLLentz) [inline lcls-plc-tmo-optics/tmo_optics/POUs/FB_AnalogInput1.TcPOU:27]: What exactly is going on here? What are these magic numbers?

File: lcls-plc-tmo-optics/tmo_optics/POUs/FB_AnalogInput1.TcPOU [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+   IF rC >= 4.0 AND rC <= 20.0 THEN
+       fReal := (fTermMax - fTermMin)*((rC - 4.0)/16) + fTermMin; // calibrate the real flow rate basic on the current calibrate from the raw bit
+   ELSE
+      fReal := 0.0;
+   END_IF
+   ]]></ST>
```
Comment (ZLLentz) [inline lcls-plc-tmo-optics/tmo_optics/POUs/FB_AnalogInput1.TcPOU:33]: After reading this, it's not clear to me why the original FB couldn't have been used (just changing the arguments)

**Review by ZLLentz — COMMENTED** [review]
I'm kind of mostly confused here, I don't understand the new FB

**Review by ZLLentz — CHANGES_REQUESTED** [review]
In the original FB the math basically just maps the output range of the sensor to the bit range in a very straightforward way. This FB seems to cannibalize the original in ways that seem unnecessary.

I don't think this is the right solution to the problem, unless there's something I don't understand about this sensor that makes it special- in which case, this should be a completely new FB instead of trying to reuse the guts of FB_AnalogInput

Does FB_AnalogInput with the following inputs not produce the right output?

- iTermBits := 15
- fTermMax := 5
- fTermMin := 0

File: lcls-plc-tmo-optics/TLR_BI_01026916.TCLRS
[the remark this answers is not in the document]
Comment (tongju12) [reply lcls-plc-tmo-optics/TLR_BI_01026916.TCLRS:1]: remove them already

File: lcls-plc-tmo-optics/TLR_BI_01026916.tclrq
[the remark this answers is not in the document]
Comment (tongju12) [reply lcls-plc-tmo-optics/TLR_BI_01026916.tclrq:1]: remove it

**Comment by tongju12** [comment]
Remove the new FB, keep the old one.  Change the iTermBits from 12 to 15 and get reasonable output as 1.85 lpm.

**Review by ZLLentz — APPROVED** [review]

**Comment by nrwslac** [comment]
Where did we get the 15 bits from? The documentation I see for the EL3054 says 12 bits. @tongju12

**Comment by ZLLentz** [comment]
The EL3054 has a 12 bit resolution, but packs the data into a 16-bit unsigned integer, using only the positive side. This is in the datasheet as `Resolution	12 bit (16 bit presentation incl. sign)` The 3 least-significant bits are essentially useless noise but the output still needs to be scaled to 15 bits.

**Comment by nrwslac** [comment]
I see, I guess this is a result of needing to represent the value using typical data types. The 12 bit must be the actual hardware bits available.

**Comment by ZLLentz** [comment]
I think you're right- this is one of the choices you can make when you're packing a 12 bit integer into a 16 bit space. (It's not the choice that I would have made, but I'm sure there were good reasons for it)

**Comment by klauer** [comment]
Ref: https://control.com/forums/threads/12-bit-ad-value-converted-to-16-bit-holding-register.6707/

**Comment by nrwslac** [comment]
Interesting, never actually considered this kind of problem before.

**Pull request merged**